src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/RabinKarp.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// An implementation of the Rabin-Karp algorithm we use as a fallback for
+    /// short inputs that we can't handle with Teddy.
+    /// https://en.wikipedia.org/wiki/Rabin%E2%80%93Karp_algorithm
+    /// </summary>
+    internal readonly struct RabinKarp
+    {
+        // The number of values we'll accept before falling back to Aho-Corasick.
+        // This also affects when Teddy may be used.
+        public const int MaxValues = 80;
+
+        // This is a tradeoff between memory consumption and the number of false positives
+        // we have to rule out during the verification step.
+        private const nuint BucketCount = 64;
+
+        // 18 = Vector128<byte>.Count + 2 (MatchStartOffset for N=3)
+        // The logic in this class is not safe from overflows, but we avoid any issues by
+        // only calling into it for inputs that are too short for Teddy to handle.
+        private const int MaxInputLength = 18 - 1;
+
+        // We're using nuint as the rolling hash, so we can spread the hash over more bits on 64bit.
+        private static int HashShiftPerElement => IntPtr.Size == 8 ? 2 : 1;
+
+        private readonly string[][] _buckets;
+        private readonly int _hashLength;
+        private readonly nuint _hashUpdateMultiplier;
+
+        public RabinKarp(ReadOnlySpan<string> values)
+        {
+            Debug.Assert(values.Length <= MaxValues);
+
+            int minimumLength = int.MaxValue;
+            foreach (string value in values)
+            {
+                minimumLength = Math.Min(minimumLength, value.Length);
+            }
+
+            Debug.Assert(minimumLength > 1);
+
+            _hashLength = minimumLength;
+            _hashUpdateMultiplier = (nuint)1 << ((minimumLength - 1) * HashShiftPerElement);
+
+            var bucketLists = new List<string>?[BucketCount];
+
+            foreach (string value in values)
+            {
+                nuint hash = 0;
+                for (int i = 0; i < minimumLength; i++)
+                {
+                    hash = (hash << HashShiftPerElement) + value[i];
+                }
+
+                nuint bucket = hash % BucketCount;
+                var bucketList = bucketLists[bucket] ??= new List<string>();
```

Comment:
There's a fair amount of allocation happening in this ctor: a 64-`List<string>` array, up to 64 `List<string>`s, a new resulting array for each created list, etc.  If there are only at most 80 values, seems like we could be leaner in various cases here, e.g. we expect very few collisions, so we could for example just store arrays of exactly the right size rather than lists, e.g. allocate a string[1] for the first string, grow it to a string[2] when there's a collision, etc.  Then we also don't need a separate bucketLists.

Reply:
Updated this as suggested